src/context.ts

```diff
@@ -134,7 +134,11 @@ export class Context {
 
   async run(tool: Tool, params: Record<string, unknown> | undefined) {
     // Tab management is done outside of the action() call.
-    const toolResult = await tool.handle(this, tool.schema.inputSchema.parse(params || {}));
+    const toolResult = await tool.handle(this, tool.schema.inputSchema.parse(params || {})).catch(e => {
+      if (e instanceof Error && /locator._generateLocatorString: Timeout .* exceeded/.test(e.message))
+        throw new Error('Ref not found. Make sure the element is present on the page before using it in the tool.');
```

Comment:
Some suggestion for inspiration, to make it more action-oriented.

```suggestion
        throw new Error('Element with this ref has been remove. Take a new snapshot to see the elements currently present on the page.');
```